### PR TITLE
C++: Better tracking of SSA memory accesses

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
This change fixes a few key problems with the existing SSA implementations:

For unaliased SSA, we were incorrectly choosing to model a local variable that had accesses that did not cover the entire variable. This has been changed to ensure that all accesses to the variable are at offset zero and have the same type as the variable itself. This was only possible to fix now that every `MemoryOperand` has its own type.

For aliased SSA, we now correctly track the offset and size of each memory access using an interval of bit offsets covered by the access. The offset interval makes the overlap computation more straightforward. Again, this is only possible now that operands have types.
The `getXXXMemoryAccess` predicates are now driven by the `MemoryAccessKind` on the operands and results, instead of by specific opcodes.

This change does fix an existing false negative in the IR dataflow tests.

I added a few simple test cases to the SSA IR tests, covering the various kinds of overlap (MustExcactly, MustTotally, and MayPartially).

I added "PrintSSA.qll", which can dump the SSA memory accesses as part of an IR dump.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SimpleSSA.qll [hunk 1]
```diff
@@ -1,12 +1,46 @@
...
+}
+
+/**
+ * Holds if the specified variable should be modeled in SSA form. For unaliased SSA, we only model a variable if its
+ * address never escapes and all reads and writes of that variable access the entire variable using the original type
+ * of the variable.
+ */
+private predicate isVariableModeled(IRVariable var) {
+  not variableAddressEscapes(var) and
+  // There's no need to check for the right size. An `IRVariable` never has an `UnknownType`, so the test for
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SimpleSSA.qll]: What happens if the AST has an `UnknownType` for a variable? Is that case excluded in IR generation?

Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SimpleSSA.qll]: I'd consider that an extractor bug. I'd prefer to fix or workaround those bugs when discovered. The enum size workaround in this PR and the decayed/inferred type fix from my previous PR are both instances of similar bugs, and both were caught by the IR sanity queries.

**Comment by rdmarsh2** [comment]
ir_gvn.ql is failing; I'm not sure yet if it just needs output accepted or if we should modify the GVN library to account for this change

**Comment by dave-bartolomeo** [comment]
Looks like my changes fixed some memory operands that were originally unmodeled, but are now modeled correctly. I've updated the test expectations.

**Pull request merged**